# Multiple Linear Regression
1. Multiple Linear Regression is an extension of Simple Linear regression as it takes more than one predictor variable to predict the response variable.
2. Multiple Linear Regression is one of the important regression algorithms which models the linear relationship between a single dependent continuous variable and more than one independent variable.
3. For MLR, the dependent or target variable(Y) must be the continuous/real, but the predictor or independent variable may be of continuous or categorical form. Each feature variable must model the linear relationship with 
the dependent variable. MLR tries to fit a regression line through a multidimensional space of data-points.

# Multiple Linear Regression  Problem Statement No. 3	
Problem Statements : 
An online car sales platform would like to improve its customer base and their experience by providing them an easy way to buy and sell cars. For this, they would like an automated model which can predict the price of the car once the user inputs the required factors. Help the business achieve their objective by applying multilinear regression on the given dataset. Please use the below columns for the analysis purpose: price, age_08_04, KM, HP, cc, Doors, Gears, Quarterly_Tax, and Weight.

Business Objectives : To predict the price of a car based on there  price, age_08_04, KM, HP, cc, Doors, Gears, Quarterly_Tax, and Weight. these parameter.

In [1]:
#lets import the libraries
import numpy as np # for mathematical and logical operations and consisting of multidimensional array objects 
import matplotlib.pyplot as plt
import pandas as pd  # for Data Manipulation
import seaborn as sns   #for Advanced visualizations
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score
from collections import Counter
from IPython.core.display import display, HTML
sns.set_style('darkgrid')

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_10312\2257052283.py:9: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


In [6]:
#lets load the data set
online = pd.read_csv("ToyotaCorolla.csv", encoding='utf-8' ,sep=",")
#as the datais not loading properly from the pd function
#we have imported the data from variable explorer

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa0 in position 0: invalid start byte

In [5]:
help(pd.read_csv)

Help on function read_csv in module pandas.io.parsers.readers:

read_csv(filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]', *, sep: 'str | None | lib.NoDefault' = <no_default>, delimiter: 'str | None | lib.NoDefault' = None, header: "int | Sequence[int] | None | Literal['infer']" = 'infer', names: 'Sequence[Hashable] | None | lib.NoDefault' = <no_default>, index_col: 'IndexLabel | Literal[False] | None' = None, usecols: 'UsecolsArgType' = None, dtype: 'DtypeArg | None' = None, engine: 'CSVEngine | None' = None, converters: 'Mapping[Hashable, Callable] | None' = None, true_values: 'list | None' = None, false_values: 'list | None' = None, skipinitialspace: 'bool' = False, skiprows: 'list[int] | int | Callable[[Hashable], bool] | None' = None, skipfooter: 'int' = 0, nrows: 'int | None' = None, na_values: 'Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None' = None, keep_default_na: 'bool' = True, na_filter: 'bool' = True, verbose: 'bool | 

In [12]:
col = online.columns.tolist()

In [14]:
print(col)

['Id', 'Model', 'Price', 'Age_08_04', 'Mfg_Month', 'Mfg_Year', 'KM', 'Fuel_Type', 'HP', 'Met_Color', 'Color', 'Automatic', 'cc', 'Doors', 'Cylinders', 'Gears', 'Quarterly_Tax', 'Weight', 'Mfr_Guarantee', 'BOVAG_Guarantee', 'Guarantee_Period', 'ABS', 'Airbag_1', 'Airbag_2', 'Airco', 'Automatic_airco', 'Boardcomputer', 'CD_Player', 'Central_Lock', 'Powered_Windows', 'Power_Steering', 'Radio', 'Mistlamps', 'Sport_Model', 'Backseat_Divider', 'Metallic_Rim', 'Radio_cassette', 'Tow_Bar']


In [6]:
online

Id                                              Model  Price  \
0        1      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13500   
1        2      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13750   
2        3      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13950   
3        4      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  14950   
4        5        TOYOTA Corolla 2.0 D4D HATCHB SOL 2/3-Doors  13750   
...    ...                                                ...    ...   
1431  1438         TOYOTA Corolla 1.3 16V HATCHB G6 2/3-Doors   7500   
1432  1439  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...  10845   
1433  1440  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...   8500   
1434  1441  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...   7250   
1435  1442        TOYOTA Corolla 1.6 LB LINEA TERRA 4/5-Doors   6950   

      Age_08_04  Mfg_Month  Mfg_Year     KM Fuel_Type   HP  Met_Color  ...  \
0            23         10      2002  46986    Diesel   90          1  ...   
1            23         10      2002  72937    Diesel   90          1  ...   
2            24          9      2002  41711    Diesel   90          1  ...   
3            26          7      2002  48000    Diesel   90          0  ...   
4            30          3      2002  38500    Diesel   90          0  ...   
...         ...        ...       ...    ...       ...  ...        ...  ...   
1431         69         12      1998  20544    Petrol   86          1  ...   
1432         72          9      1998  19000    Petrol   86          0  ...   
1433         71         10      1998  17016    Petrol   86          0  ...   
1434         70         11      1998  16916    Petrol   86          1  ...   
1435         76          5      1998      1    Petrol  110          0  ...   

     Central_Lock  Powered_Windows  Power_Steering  Radio  Mistlamps  \
0               1                1               1      0          0   
1               1                0               1      0          0   
2               0                0               1      0          0   
3               0                0               1      0          0   
4               1                1               1      0          1   
...           ...              ...             ...    ...        ...   
1431            1                1               1      0          1   
1432            0                0               1      0          0   
1433            0                0               1      0          0   
1434            0                0               0      0          0   
1435            0                0               1      0          0   

      Sport_Model  Backseat_Divider  Metallic_Rim  Radio_cassette  Tow_Bar  
0               0                 1             0               0        0  
1               0                 1             0               0        0  
2               0                 1             0               0        0  
3               0                 1             0               0        0  
4               0                 1             0               0        0  
...           ...               ...           ...             ...      ...  
1431            1                 1             0               0        0  
1432            1                 1             0               0        0  
1433            0                 1             0               0        0  
1434            0                 1             0               0        0  
1435            0                 0             0               0        0  

[1436 rows x 38 columns]

In [7]:
online.shape

(1436, 38)

In [10]:
online.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 38 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Id                1436 non-null   int64 
 1   Model             1436 non-null   object
 2   Price             1436 non-null   int64 
 3   Age_08_04         1436 non-null   int64 
 4   Mfg_Month         1436 non-null   int64 
 5   Mfg_Year          1436 non-null   int64 
 6   KM                1436 non-null   int64 
 7   Fuel_Type         1436 non-null   object
 8   HP                1436 non-null   int64 
 9   Met_Color         1436 non-null   int64 
 10  Color             1436 non-null   object
 11  Automatic         1436 non-null   int64 
 12  cc                1436 non-null   int64 
 13  Doors             1436 non-null   int64 
 14  Cylinders         1436 non-null   int64 
 15  Gears             1436 non-null   int64 
 16  Quarterly_Tax     1436 non-null   int64 
 17  Weight        

In [11]:
online.isnull().sum()

Id                  0
Model               0
Price               0
Age_08_04           0
Mfg_Month           0
Mfg_Year            0
KM                  0
Fuel_Type           0
HP                  0
Met_Color           0
Color               0
Automatic           0
cc                  0
Doors               0
Cylinders           0
Gears               0
Quarterly_Tax       0
Weight              0
Mfr_Guarantee       0
BOVAG_Guarantee     0
Guarantee_Period    0
ABS                 0
Airbag_1            0
Airbag_2            0
Airco               0
Automatic_airco     0
Boardcomputer       0
CD_Player           0
Central_Lock        0
Powered_Windows     0
Power_Steering      0
Radio               0
Mistlamps           0
Sport_Model         0
Backseat_Divider    0
Metallic_Rim        0
Radio_cassette      0
Tow_Bar             0
dtype: int64

In [9]:
online.describe()

Id         Price    Age_08_04    Mfg_Month     Mfg_Year  \
count  1436.000000   1436.000000  1436.000000  1436.000000  1436.000000   
mean    721.555014  10730.824513    55.947075     5.548747  1999.625348   
std     416.476890   3626.964585    18.599988     3.354085     1.540722   
min       1.000000   4350.000000     1.000000     1.000000  1998.000000   
25%     361.750000   8450.000000    44.000000     3.000000  1998.000000   
50%     721.500000   9900.000000    61.000000     5.000000  1999.000000   
75%    1081.250000  11950.000000    70.000000     8.000000  2001.000000   
max    1442.000000  32500.000000    80.000000    12.000000  2004.000000   

                  KM           HP    Met_Color    Automatic           cc  ...  \
count    1436.000000  1436.000000  1436.000000  1436.000000   1436.00000  ...   
mean    68533.259749   101.502089     0.674791     0.055710   1576.85585  ...   
std     37506.448872    14.981080     0.468616     0.229441    424.38677  ...   
min         1.000000    69.000000     0.000000     0.000000   1300.00000  ...   
25%     43000.000000    90.000000     0.000000     0.000000   1400.00000  ...   
50%     63389.500000   110.000000     1.000000     0.000000   1600.00000  ...   
75%     87020.750000   110.000000     1.000000     0.000000   1600.00000  ...   
max    243000.000000   192.000000     1.000000     1.000000  16000.00000  ...   

       Central_Lock  Powered_Windows  Power_Steering        Radio  \
count   1436.000000      1436.000000     1436.000000  1436.000000   
mean       0.580084         0.561978        0.977716     0.146240   
std        0.493717         0.496317        0.147657     0.353469   
min        0.000000         0.000000        0.000000     0.000000   
25%        0.000000         0.000000        1.000000     0.000000   
50%        1.000000         1.000000        1.000000     0.000000   
75%        1.000000         1.000000        1.000000     0.000000   
max        1.000000         1.000000        1.000000     1.000000   

         Mistlamps  Sport_Model  Backseat_Divider  Metallic_Rim  \
count  1436.000000  1436.000000       1436.000000   1436.000000   
mean      0.256964     0.300139          0.770195      0.204735   
std       0.437111     0.458478          0.420854      0.403649   
min       0.000000     0.000000          0.000000      0.000000   
25%       0.000000     0.000000          1.000000      0.000000   
50%       0.000000     0.000000          1.000000      0.000000   
75%       1.000000     1.000000          1.000000      0.000000   
max       1.000000     1.000000          1.000000      1.000000   

       Radio_cassette      Tow_Bar  
count     1436.000000  1436.000000  
mean         0.145543     0.277855  
std          0.352770     0.448098  
min          0.000000     0.000000  
25%          0.000000     0.000000  
50%          0.000000     0.000000  
75%          0.000000     1.000000  
max          1.000000     1.000000  

[8 rows x 35 columns]

# Data Preprocessing

In [24]:
# Checking the null values, count, columns name and  data describetion  
online.describe()
online.columns
online.count()
online.isna().sum()
online.info

<bound method DataFrame.info of         Id                                              Model  Price  \
0        1      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13500   
1        2      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13750   
2        3     �TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  13950   
3        4      TOYOTA Corolla 2.0 D4D HATCHB TERRA 2/3-Doors  14950   
4        5        TOYOTA Corolla 2.0 D4D HATCHB SOL 2/3-Doors  13750   
...    ...                                                ...    ...   
1431  1438         TOYOTA Corolla 1.3 16V HATCHB G6 2/3-Doors   7500   
1432  1439  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...  10845   
1433  1440  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...   8500   
1434  1441  TOYOTA Corolla 1.3 16V HATCHB LINEA TERRA 2/3-...   7250   
1435  1442        TOYOTA Corolla 1.6 LB LINEA TERRA 4/5-Doors   6950   

      Age_08_04  Mfg_Month  Mfg_Year     KM Fuel_Type   HP  Met_Color  ...  \
0            23         1

In [26]:
#as the first column is given serial number, so this is not required
online = online.drop('Id', axis=1)
online = online.drop('Model',axis=1)
online

Price  Age_08_04  Mfg_Month  Mfg_Year     KM Fuel_Type   HP  Met_Color  \
0     13500         23         10      2002  46986    Diesel   90          1   
1     13750         23         10      2002  72937    Diesel   90          1   
2     13950         24          9      2002  41711    Diesel   90          1   
3     14950         26          7      2002  48000    Diesel   90          0   
4     13750         30          3      2002  38500    Diesel   90          0   
...     ...        ...        ...       ...    ...       ...  ...        ...   
1431   7500         69         12      1998  20544    Petrol   86          1   
1432  10845         72          9      1998  19000    Petrol   86          0   
1433   8500         71         10      1998  17016    Petrol   86          0   
1434   7250         70         11      1998  16916    Petrol   86          1   
1435   6950         76          5      1998      1    Petrol  110          0   

       Color  Automatic  ...  Central_Lock  Powered_Windows  Power_Steering  \
0       Blue          0  ...             1                1               1   
1     Silver          0  ...             1                0               1   
2       Blue          0  ...             0                0               1   
3      Black          0  ...             0                0               1   
4      Black          0  ...             1                1               1   
...      ...        ...  ...           ...              ...             ...   
1431    Blue          0  ...             1                1               1   
1432    Grey          0  ...             0                0               1   
1433    Blue          0  ...             0                0               1   
1434    Grey          0  ...             0                0               0   
1435   Green          0  ...             0                0               1   

      Radio  Mistlamps  Sport_Model  Backseat_Divider  Metallic_Rim  \
0         0          0            0                 1             0   
1         0          0            0                 1             0   
2         0          0            0                 1             0   
3         0          0            0                 1             0   
4         0          1            0                 1             0   
...     ...        ...          ...               ...           ...   
1431      0          1            1                 1             0   
1432      0          0            1                 1             0   
1433      0          0            0                 1             0   
1434      0          0            0                 1             0   
1435      0          0            0                 0             0   

      Radio_cassette  Tow_Bar  
0                  0        0  
1                  0        0  
2                  0        0  
3                  0        0  
4                  0        0  
...              ...      ...  
1431               0        0  
1432               0        0  
1433               0        0  
1434               0        0  
1435               0        0  

[1436 rows x 36 columns]

In [27]:
#lets make the dependant and independant variables
#as from the data set profit is depends on the remaining all
X = online.iloc[:,1:36] #independ variables excluding price 
= online.iloc[:,0]  #depended variable price

In [28]:
#as the State column is catergorical we will convert this using onehot encoding using get_dummies()
#the impact of first NewYork is removed
Fuel_Type= pd.get_dummies(X['Fuel_Type'])
Color = pd.get_dummies(X['Color'])

In [29]:
#lets remove the categorical column
X = X.drop(['Color','Fuel_Type'], axis=1)
X 

Age_08_04  Mfg_Month  Mfg_Year     KM   HP  Met_Color  Automatic    cc  \
0            23         10      2002  46986   90          1          0  2000   
1            23         10      2002  72937   90          1          0  2000   
2            24          9      2002  41711   90          1          0  2000   
3            26          7      2002  48000   90          0          0  2000   
4            30          3      2002  38500   90          0          0  2000   
...         ...        ...       ...    ...  ...        ...        ...   ...   
1431         69         12      1998  20544   86          1          0  1300   
1432         72          9      1998  19000   86          0          0  1300   
1433         71         10      1998  17016   86          0          0  1300   
1434         70         11      1998  16916   86          1          0  1300   
1435         76          5      1998      1  110          0          0  1600   

      Doors  Cylinders  ...  Central_Lock  Powered_Windows  Power_Steering  \
0         3          4  ...             1                1               1   
1         3          4  ...             1                0               1   
2         3          4  ...             0                0               1   
3         3          4  ...             0                0               1   
4         3          4  ...             1                1               1   
...     ...        ...  ...           ...              ...             ...   
1431      3          4  ...             1                1               1   
1432      3          4  ...             0                0               1   
1433      3          4  ...             0                0               1   
1434      3          4  ...             0                0               0   
1435      5          4  ...             0                0               1   

      Radio  Mistlamps  Sport_Model  Backseat_Divider  Metallic_Rim  \
0         0          0            0                 1             0   
1         0          0            0                 1             0   
2         0          0            0                 1             0   
3         0          0            0                 1             0   
4         0          1            0                 1             0   
...     ...        ...          ...               ...           ...   
1431      0          1            1                 1             0   
1432      0          0            1                 1             0   
1433      0          0            0                 1             0   
1434      0          0            0                 1             0   
1435      0          0            0                 0             0   

      Radio_cassette  Tow_Bar  
0                  0        0  
1                  0        0  
2                  0        0  
3                  0        0  
4                  0        0  
...              ...      ...  
1431               0        0  
1432               0        0  
1433               0        0  
1434               0        0  
1435               0        0  

[1436 rows x 33 columns]

In [30]:
#lets use concatination
X = pd.concat([X,Color,Fuel_Type], axis=1)
X 

Age_08_04  Mfg_Month  Mfg_Year     KM   HP  Met_Color  Automatic    cc  \
0            23         10      2002  46986   90          1          0  2000   
1            23         10      2002  72937   90          1          0  2000   
2            24          9      2002  41711   90          1          0  2000   
3            26          7      2002  48000   90          0          0  2000   
4            30          3      2002  38500   90          0          0  2000   
...         ...        ...       ...    ...  ...        ...        ...   ...   
1431         69         12      1998  20544   86          1          0  1300   
1432         72          9      1998  19000   86          0          0  1300   
1433         71         10      1998  17016   86          0          0  1300   
1434         70         11      1998  16916   86          1          0  1300   
1435         76          5      1998      1  110          0          0  1600   

      Doors  Cylinders  ...  Green  Grey  Red  Silver  Violet  White  Yellow  \
0         3          4  ...      0     0    0       0       0      0       0   
1         3          4  ...      0     0    0       1       0      0       0   
2         3          4  ...      0     0    0       0       0      0       0   
3         3          4  ...      0     0    0       0       0      0       0   
4         3          4  ...      0     0    0       0       0      0       0   
...     ...        ...  ...    ...   ...  ...     ...     ...    ...     ...   
1431      3          4  ...      0     0    0       0       0      0       0   
1432      3          4  ...      0     1    0       0       0      0       0   
1433      3          4  ...      0     0    0       0       0      0       0   
1434      3          4  ...      0     1    0       0       0      0       0   
1435      5          4  ...      1     0    0       0       0      0       0   

      CNG  Diesel  Petrol  
0       0       1       0  
1       0       1       0  
2       0       1       0  
3       0       1       0  
4       0       1       0  
...   ...     ...     ...  
1431    0       0       1  
1432    0       0       1  
1433    0       0       1  
1434    0       0       1  
1435    0       0       1  

[1436 rows x 46 columns]

In [31]:
#let's split the data into train and test
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [32]:
#now lets build a model using fitting Multilinear Regression to the training set
from sklearn.linear_model import LinearRegression
start_f = LinearRegression() 
start_f.fit(X_train, y_train)

LinearRegression()

In [33]:
#prediction for the test results
y_pred = start_f.predict(X_test)
y_pred

array([  8902.89094415,  10930.17351334,  13378.01869454,   7677.83247574,
        10432.66576372,   9915.24818761,   6462.49542937,  15332.47255526,
        11479.46751546,  10619.19257447,   6377.88558016,  10074.68400938,
        11118.36064289,   9713.68177557,   7302.89934513,  14768.94772503,
        11436.06948335,   8115.18617751,  11504.16805212,  10862.43148416,
         6885.94236157,   6653.12150519,  11508.87802906,  11798.71454718,
        21631.77943184,   9198.01428745,  13466.87408117,   6033.73766786,
         9283.95849611,   7675.30542709,  10994.43115259,   8381.84334864,
        16348.52189393,   7465.00611289,   9671.861632  ,  11876.21165276,
         9763.45479031,  12415.93153024,  10009.73675029,   8887.69810875,
         8356.28868037,   9952.18199699,  10277.85282611,  15242.64858869,
         9943.24673131,   7259.90101591,  10095.56221832,   6849.91378228,
         8253.41964331,   8672.59746076,  20915.69026066,   9161.48285269,
         7445.49539548,  

In [34]:
#for comarision to check whether the y_test is good or y_pred is good
#we will use r^2 score from the sklearn package
from sklearn.metrics import r2_score
score = r2_score(y_test, y_pred) 
score

0.37407949610941227

In [17]:
# Predicting Cross Validation Score the Test set results
cv_linear = cross_val_score(estimator = start_f, X = X_train, y = y_train, cv = 10)
cv_linear

array([0.85295169, 0.90241947, 0.90174145, 0.91481452, 0.92485328,
       0.9108801 , 0.91970749, 0.82267693, 0.92849608, 0.94058836])

In [18]:
# Predicting R2 Score the Train set results
y_pred_linear_train = start_f.predict(X_train)
r2_score_linear_train = r2_score(y_train, y_pred_linear_train)
r2_score_linear_train

0.9207693782607749

In [19]:
# Predicting R2 Score the Test set results
y_pred_linear_test = start_f.predict(X_test)
r2_score_linear_test = r2_score(y_test, y_pred_linear_test)
r2_score_linear_test

0.37407949610941227

In [20]:
# Predicting RMSE the Test set results
rmse_linear = (np.sqrt(mean_squared_error(y_test, y_pred_linear_test)))
print("CV: ", cv_linear.mean())
print('R2_score (train): ', r2_score_linear_train)
print('R2_score (test): ', r2_score_linear_test)
print("RMSE: ", rmse_linear)

CV:  0.9019129374534772
R2_score (train):  0.9207693782607749
R2_score (test):  0.37407949610941227
RMSE:  2790.6780509855835


Result : By checking the result of model the train score is 92% while train socre is 37% mean our model is under fitting condition is occured 